src/library/scala/collection/mutable/HashMap.scala

```diff
@@ -29,133 +32,288 @@ import scala.collection.{Iterator, MapFactory, StrictOptimizedIterableOps, Stric
   *  @define mayNotTerminateInf
   *  @define willNotTerminateInf
   */
-class HashMap[K, V]
+class HashMap[K, V](initialCapacity: Int, loadFactor: Double)
```

Comment:
Can we make it final..? I guess the main use case for extending it is to override `default`. We can make that default function an assignable field (`withDefaultValue`?)

Reply:
There are 3 other subclasses in the compiler: `QuantMap` in `scala.reflect.internal.util.Statistics`, `result` in `scala.tools.nsc.doc.model.IndexModelFactory` and `ReponseMap` in `scala.tools.nsc.interactive.Global`. Plus some more in test cases. Making it final would probably break a lot of code if we already have that much in our own codebase.

Reply:
I still think we should make it final / deprecatedOverriding. The 3 examples don't do much more than adding defaults, I think people should use composition not inheritance here. Allowing subclasses of HashMap makes it hard for us to change things, to optimize the implementaiton, and even changes that look safe might break subclasses in unexpected ways - you cannot anticipate what implementation details subclasses rely on.

For convenience, we could provide a non-final `DelegatingMap` that people could extend to add methods to a map. But just doing it for `mutable.Map` seems unsystematic. We have many other final collection types in the library, I don't see why `mutable.HashMap` should be an exception.

Reply:
Actually the `default` method is not specific to HashMap, it is defined in `collection.Map` and we already have `withDefault` to create a delegate.